### PR TITLE
Fix #2456: Optimize multiplications by 0 and by powers of 2.

### LINKED ISSUE #2456: Rewrite multiplications by powers of 2 as shifts
Typical examples:
* `x * 4` -> `x << 2`
* ~~`x / 4` -> `x >> 2`~~ (division does not work, see below)

This might seem like an optimization that the JS VM could do on its own, but not quite. For multiplication of `Int`s, the JS code we have is `$imul(4, x)`, which, although `$imul` is a built-in, requires more work from the JIT generated code than a simple operator `x << 2`.

For multiplication and division of `Long`s, this is obviously a big win, as the implementations of shifts are much simpler than multiply and divide.

For division of `Int`s, it's less clear that it's a win, but it makes sense to apply the rewriting there too, if only for symmetry. It can't hurt, at least (and it requires less characters in the generated JS file anyway).

### PR DESCRIPTION
**Pull request by sjrd**
- `0 * n` -> `{ n; 0 }`
- `x * n` if `x == 2^k` -> `n << k`

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @nicolasstucki 
Test suite diff: https://gist.github.com/sjrd/53e8c486e992a95f2cc81caaecb4d569 (without the tests added by this PR)

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala [hunk 1]
```diff
@@ -2571,10 +2571,24 @@ private[optimizer] abstract class OptimizerCore(
           case (_, PreTransLit(IntLiteral(_))) =>
             foldBinaryOp(Int_*, rhs, lhs)
 
-          case (PreTransLit(IntLiteral(1)), _)  =>
-            rhs
-          case (PreTransLit(IntLiteral(-1)), _) =>
-            foldBinaryOp(Int_-, PreTransLit(IntLiteral(0)), rhs)
+          case (PreTransLit(IntLiteral(l)), _) =>
```
Comment (nicolasstucki) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala]: Consider not using lowercase L for the literal.

**Comment by nicolasstucki** [comment]
That is all. Otherwise LGTM.

**Comment by sjrd** [comment]
Updated.

**Comment by nicolasstucki** [comment]
LGTM

**Pull request merged**